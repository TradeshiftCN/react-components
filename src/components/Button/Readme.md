### Basic Usage

```jsx
<Button type="primary">Primary</Button>
<Button type="warning">Warning</Button>
<Button type="danger">Danger</Button>
<Button>Default</Button>
```

### Icons

```jsx
<Button type="primary" icon="download"></Button>
<Button icon="download">Download</Button>
<Button icon="arrowleft" label="Go back"></Button>
<Button>Go forward <i className="ts-icon-arrowright"/></Button>
```

### Size

_Note: Don't use **small** size with **icon**_

```jsx
<Button type="primary" size="small">Primary</Button>
<Button size="small">Default</Button>
```

### Loading

```jsx
initialState = { loading: false };
<Button
	type="primary"
	loading={state.loading}
	onClick={() => {
		setState({ loading: true });
		setTimeout(() => {
			setState({ loading: false });
		}, 5000);
	}}
>
	Click me
</Button>;
```
