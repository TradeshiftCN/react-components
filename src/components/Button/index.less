@import '../../style/variables/index';

@keyframes ts-rc-move-progress-bar {
	0% {
		background-position: 0 0;
	}
	100% {
		background-position: 40px 40px;
	}
}

.ts-rc-button {
	display: inline-block;
	padding-left: @ts-unit;
	padding-right: @ts-unit;
	border: 1px solid transparent;
	line-height: @ts-unit-double - 2px;
	border-radius: @ts-unit;
	transition: @ts-transition-now;
	font-size: @ts-fontsize;
	position: relative;
	&:not(:disabled):hover {
		text-decoration: none;
	}
	& + & {
		margin-left: 5px;
	}
	&--small {
		padding-left: @ts-unit-half;
		padding-right: @ts-unit-half;
		line-height: @ts-unit - 2px;
		font-size: @ts-fontsize-small;
	}
	&--default {
		border-color: @ts-color-gray-light;
		color: @ts-color-black;
		&:hover {
			background-color: @ts-color-gray-lighter;
		}
	}
	&--primary {
		background-color: @ts-color-blue;
		border-color: @ts-color-blue;
		color: @ts-color-white;
		&:hover {
			background-color: @ts-color-blue-dark;
		}
	}
	&--warning {
		background-color: @ts-color-orange;
		color: @ts-color-white;
		border-color: @ts-color-white;
	}
	&--danger {
		background-color: @ts-color-red;
		color: @ts-color-white;
		border-color: @ts-color-white;
	}
	&--circle {
		padding-left: 0;
		padding-right: 0;
		width: @ts-unit-double;
		border-radius: @ts-unit;
	}
	&--loading {
		&,
		&:hover {
			background-color: transparent;
			border-color: transparent;
			cursor: default;
		}
		& > * {
			visibility: hidden;
		}
		&::after {
			content: '';
			display: block;
			position: absolute;
			visibility: visible;
			left: 0;
			top: @ts-unit-half;
			right: 0;
			bottom: @ts-unit-half;
			background-image: linear-gradient(
				-45deg,
				hsla(198, 20%, 83%, 0.65) 25%,
				hsla(200, 21%, 92%, 0.65) 25%,
				hsla(200, 21%, 92%, 0.65) 50%,
				hsla(198, 20%, 83%, 0.65) 50%,
				hsla(198, 20%, 83%, 0.65) 75%,
				hsla(200, 21%, 92%, 0.65) 75%,
				hsla(200, 21%, 92%, 0.65)
			);
			background-size: @ts-unit-double @ts-unit-double;
			border-radius: inherit;
			animation: ts-rc-move-progress-bar 1s linear infinite;
		}
	}
	& > span {
		text-transform: uppercase;
	}
	& > i {
		font-size: @ts-fontsize-icon;
		vertical-align: middle;
	}
	& > * + * {
		margin-left: 8px;
	}
}
