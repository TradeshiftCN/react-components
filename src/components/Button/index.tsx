import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import _ from 'lodash';

export type ButtonProps = {
	type?: 'default' | 'primary' | 'danger' | 'warning';
	className?: string;
	loading?: boolean;
	size?: 'small';
	children?: React.ReactNode;
	icon?: string;
	label?: string;
	disabled?: boolean;
} & React.DetailedHTMLProps<React.HTMLAttributes<HTMLButtonElement>, HTMLButtonElement>;

const prefixCls = 'ts-rc-button';

function genButtonLabel(children: React.ReactNode) {
	return React.Children.map(children, child => {
		if (_.isString(child)) {
			return <span>{child}</span>;
		}

		return child;
	});
}

function Button(props: ButtonProps) {
	const { type, className, loading, size, children, icon, label, disabled, ...buttonProps } = props;
	const labelNodes = React.Children.count(children) ? children : label;
	const hasOnlyIcon = !_.isNil(icon) && _.isNil(labelNodes);
	if (_.isNil(labelNodes) && _.isNil(icon)) {
		return null;
	}

	if (size === 'small' && icon) {
		throw new Error("small size can't used with icon");
	}

	return (
		<button
			{...buttonProps}
			className={cx(prefixCls, className, `${prefixCls}--${type}`, {
				[`${prefixCls}--${size}`]: !!size,
				[`${prefixCls}--circle`]: hasOnlyIcon,
				[`${prefixCls}--loading`]: !!loading
			})}
			disabled={!!loading || !!props.disabled}
		>
			{!!icon && <i className={`ts-icon-${icon}`} />}
			{!_.isNil(labelNodes) && genButtonLabel(labelNodes)}
		</button>
	);
}

Button.propTypes = {
	type: PropTypes.oneOf(['default', 'primary', 'danger', 'warning']),
	loading: PropTypes.bool,
	className: PropTypes.string,
	size: PropTypes.oneOf(['small']),
	children: PropTypes.node,
	icon: PropTypes.string,
	label: PropTypes.string,
	disabled: PropTypes.bool
};

Button.defaultProps = {
	type: 'default',
	loading: false
};

export default Button;
