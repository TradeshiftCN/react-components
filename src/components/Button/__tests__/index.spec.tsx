import React from 'react';
import { shallow } from 'enzyme';
import Button from '..';

describe('Button', () => {
	it('basic', () => {
		const wrapper = shallow(<Button>btn</Button>);
		expect(wrapper).toMatchSnapshot();
		expect(wrapper.html()).toBe(shallow(<Button label="btn" />).html());

		wrapper.setProps({
			size: 'small',
			type: 'primary',
			loading: true
		});
		expect(wrapper).toMatchSnapshot();
	});

	it('icon', () => {
		const wrapper = shallow(<Button icon="test" />);
		expect(wrapper).toMatchSnapshot();

		wrapper.setProps({ label: 'text' });
		expect(wrapper).toMatchSnapshot();
	});

	it('empty', () => {
		const wrapper = shallow(<Button />);
		expect(wrapper.isEmptyRender()).toBe(true);
	});

	it('children', () => {
		const wrapper = shallow(
			<Button>
				text <b>element</b>
			</Button>
		);
		expect(wrapper).toMatchSnapshot();
	});

	it('exception', () => {
		expect(() => shallow(<Button icon="test" size="small" />)).toThrowError();
	});
});
