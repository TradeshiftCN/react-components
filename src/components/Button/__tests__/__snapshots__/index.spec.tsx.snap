// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button basic 1`] = `
<button
  className="ts-rc-button ts-rc-button--default"
  disabled={false}
>
  <span
    key=".0"
  >
    btn
  </span>
</button>
`;

exports[`Button basic 2`] = `
<button
  className="ts-rc-button ts-rc-button--primary ts-rc-button--small ts-rc-button--loading"
  disabled={true}
>
  <span
    key=".0"
  >
    btn
  </span>
</button>
`;

exports[`Button children 1`] = `
<button
  className="ts-rc-button ts-rc-button--default"
  disabled={false}
>
  <span
    key=".0"
  >
    text 
  </span>
  <b
    key=".1"
  >
    element
  </b>
</button>
`;

exports[`Button icon 1`] = `
<button
  className="ts-rc-button ts-rc-button--default ts-rc-button--circle"
  disabled={false}
>
  <i
    className="ts-icon-test"
  />
</button>
`;

exports[`Button icon 2`] = `
<button
  className="ts-rc-button ts-rc-button--default"
  disabled={false}
>
  <i
    className="ts-icon-test"
  />
  <span
    key=".0"
  >
    text
  </span>
</button>
`;
